w11h03/test/pgdp/networking/HTTPTest.java

```diff
@@ -0,0 +1,327 @@
+        assertEquals("d2hhdCBhcmUgeWEgc25vb3BpbmcgYXJvdW5kIGZvcj8=", requestToken("Alfred", "Wagner"));
+    }
+
+    @Test
+    @DisplayName("requestToken – unsuccessful")
+    void testRequestTokenFailed() {
+        client.respond(422, error);
+        assertNull(requestToken("Alfred", "Wagner"));
+    }
+
+    @Test
+    @DisplayName("requestToken – garbage data")
+    void testRequestTokenGarbage() {
+        client.respond(418, "I'm a teapot.");
+        assertNull(requestToken("Alfred", "Wagner"));
+    }
+
+    @Test
+    @DisplayName("login – success")
+    void testLoginSuccess() {
+        client
+                .respond(200, """
+                {
+                  "access_token": "YXJlIHlvdSBsb29raW5nIGZvciBzb21ldGhpbmc/",
+                  "token_type": "irrelevant"
+                }""")
+                .assertNextRequest(httpRequest -> {
+                    var authorization = httpRequest.headers().allValues("Authorization");
+                    assertNotNull(authorization);
+                    assertTrue(authorization.size() > 0);
+                    assertEquals("Bearer YXJlIHlvdSBsb29raW5nIGZvciBzb21ldGhpbmc/", authorization.get(0));
+                })
+                .respond(200, """
+                {
+                  "username": "Alfred",
+                  "id": 439287
+                }""");
+        assertTrue(dataHandler.login("Alfred", "Wagner"));
+        assertEquals("Alfred", getField(dataHandler, "username"));
+        assertEquals("Wagner", getField(dataHandler, "password"));
+        assertEquals(439287, getFieldInt(dataHandler, "id"));
+    }
+
+    @Test
+    @DisplayName("login – no token")
+    void testLoginTokenFailed() {
+        client
+                .respond(422,  error);
+        assertFalse(dataHandler.login("Alfred", "Wagner"));
+        assertNull(getField(dataHandler, "username"));
+        assertNull(getField(dataHandler, "password"));
+        assertEquals(0, getFieldInt(dataHandler, "id"));
+    }
+
+    @Test
+    @DisplayName("login – unsuccessful")
+    void testLoginFailed() {
+        client
+                .respond(200, """
+                {
+                  "access_token": "aG93IGxvbmcgYXJlIHlvdSBnb25uYSBrZWVwIGRvaW5nIHRoaXM=",
+                  "token_type": "irrelevant"
+                }""")
+                .assertNextRequest(httpRequest -> {
+                    var authorization = httpRequest.headers().allValues("Authorization");
+                    assertNotNull(authorization);
+                    assertTrue(authorization.size() > 0);
+                    assertEquals("Bearer aG93IGxvbmcgYXJlIHlvdSBnb25uYSBrZWVwIGRvaW5nIHRoaXM=", authorization.get(0));
+                })
+                .respond(422, error);
+        assertFalse(dataHandler.login("Alfred", "Wagner"));
+        assertEquals("Alfred", getField(dataHandler, "username"));
+        assertEquals("Wagner", getField(dataHandler, "password"));
+        assertEquals(0, getFieldInt(dataHandler, "id"));
```

Comment:
Should we set `username`, `password` and `id` even when the login failed?

Reply:
From understanding is "Falls diese Anfrage erfolgreich ist" when the login was successful. However, in this case the login wasn't successful.

I have asked on Zulip :) https://zulip.in.tum.de/#narrow/stream/1525-PGdP-W11H03/topic/login.20Methode.20Attribute/near/909604

Reply:
Ah neat. Thank you :)

I was relatively sure that meant something else and really didn't want to bother the ÜL more than I already have.

Reply:
> Wenn !login() -> keine attribute gesetzt

https://zulip.in.tum.de/#narrow/stream/1525-PGdP-W11H03/topic/login.20Methode.20Attribute/near/909609

So when the login was unsuccessful should check if the attributes are `null`